Introducing yaaf (#98)
rootstrap/blog: content/Hello yaaf - Juan Manuel Ramallo.md

> ## Yet another active form

Author: ```suggestion
## Do you use form objects in your Rails apps?
```

Author: I find the title "Yet another active form" to be not entirely expressive or catchy at all, what do you think of changing the title to the suggestion above? @andrea-fajardo

> ## Yet another active form

Reviewer: Yet Another Active Form | Using form objects in Rails Apps

> ## Yet another active form

Reviewer 2: ```
## Yet Another Active Form | Using form objects in Rails Apps
```

> ## Yet another active form

Reviewer 2: ## Yet Another Active Form | Using form objects in Rails Apps
```

> ## Yet another active form

Reviewer 2: ```suggestion
## Yet another active form
```

> Introducing [yaaf](https://github.com/rootstrap/yaaf), a gem to ease the usage of the form object pattern in rails apps.

Reviewer 2: ```suggestion
Introducing [yaaf](https://github.com/rootstrap/yaaf), a gem to ease the usage of the form object pattern in rails apps.
```

> Introducing [yaaf](https://github.com/rootstrap/yaaf), a gem to ease the usage of the form object pattern in rails apps.

Reviewer: Introducing Yet Another Active Form [yaaf]...

> Introducing [yaaf](https://github.com/rootstrap/yaaf), a gem to ease the usage of the form object pattern in rails apps.

Reviewer 2: ```suggestion
Introducing Yet Another Active Form [yaaf](https://github.com/rootstrap/yaaf), a gem to ease the usage of the form object pattern in rails apps.
```

> ### ~The~ My pain points

Reviewer 3: This section is titled _My pain points_, however in the content you are always referring to _we_. I think the title implies that this is some sort of personal story, however you are speaking in plural. What about changing the title to _Our pain points_ or changing the redaction to a first person style?

Author: That is totally a good point. The thing is want to share what I consider pain point while working with form objects. 

In the paragraph when I say "we", I refer to the rails community:
```txt
The form object pattern is widely used across Rails apps,
and yet we tend to write these form objects [...]
```
I would rather change the paragraph with "I" instead of "we" if you prefer that, but I'd rather leave it as it is. What do you think?

Reviewer 3: Okay, I see. It's fine by me then.

> ### ~The~ My pain points

Reviewer: Form Object Pain Points

> ### ~The~ My pain points

Reviewer 2: ```suggestion
### Form Object Pain Points
```

The form object pattern is widely used across Rails apps,

> and yet we tend to write these form objects in a different manner from project to project.

Reviewer 3: What do you think about using the word way instead of manner here. Something in the lines of: _we tend to write these form objects in different ways from project to project_.

Author: Oh I like the word "manner" more than "way", I find it a bit more sophisticated, fancier than "way"

> and yet we tend to write these form objects in a different manner from project to project.

Reviewer 2: ```suggestion
and yet we tend to write these form objects in a different ways from project to project.
```

Most of the time, they don't even have the same interface across form objects in the same app.

We also tend to write validations twice, in the model and in the form object,
just so that the form object has the errors and the view can display them.

> Well, I can say that most of our in-house form objects make well use of database transactions.

Reviewer 3: Here I would change the word well for _good_

Author: I don't think they mean the same thing. With "Well" I meant to create a pause before agreeing with the reader on a matter, the matter is that "all existing form objects are using database transactions correctly"

Reviewer 4: I don't get the pause sense there. I agree with @juan-apa, `well` sounds odd for me too.

Author: Ok, two disagreements. I'll change. 

What do you think of removing it altogether? 
```suggestion
I can say that most of our in-house form objects make well use of database transactions.
```

Reviewer 5: I believe they meant to change the "make _well_ use of" to "make _good_ use of", or at least thats what I would change here. 😆I do get what you want to achieve with the first well.

Author: hahahahha good catch nacho! thanks!

```suggestion
Well, I can say that most of our in-house form objects make good use of database transactions.
```
Like this then?

But I bet we rarely provide the controller the ability to use a banged method,
expecting it to raise an error if things go south.

Business logic is present everywhere in your app, controllers, models, helpers (some hardcore scenarios might
include Rails initializers), but is that right? Many times we find ourselves with the need to add business logic
around the creation of an object, such as sending an email, updating other records or even calling an external service.
We know that the controller is not a good place, and the model would be terrible as well because it violates SRP.
So we end up creating several service objects and throwing the business logic there as if it were a trashcan.

### Instead of all of that, we can...

> use yaaf. yaaf solves all of that in an easy way, let me explain:

Reviewer 4: I would change this to `use yaaf, as it solves all of that in an easy way. Let me explain:` so that you don't repeat `yaaf` immediately after.

Author: Cool, what do you think of:
```suggestion
use yaaf, it solves all of that in an easy way, let me explain:
```

- Same interface? yaaf will encourage you to use `save` and just `save` to, well you know,
SAVE the data from your form into models.

> - Contextual validations only? Please do that, and keep data integrity validations where they belong,

Reviewer 6: Not sure if it's clear that contextual validations should be in the form objects.. alternative:

```suggestion
- Contextual validations only? Please put them in your form objects, and keep data integrity validations where they belong,
```

> in your service objects, haha no, of course no, in your models!

Reviewer 3: I think the correct word here is _not_ (where it says _"of course no"_)

Author: Definitely! 
```suggestion
in your service objects, haha no, of course not, those belong in your models!
```
How does that look?

Reviewer 4: Is this on the next line ?

Author: In markdown yeah, but in the output they will be in the same line

yaaf will take the liberty to promote all model errors collections into the form object errors collection,
so you can show them in your views

> - So one the form objects need to raise an error when not saved? As you may have guessed already,

Reviewer 7: Missing a comma after So

Author: ```suggestion
- So, one the form objects need to raise an error when not saved? As you may have guessed already,
```
Like that?

Author: ```suggestion
- So, one of the form objects need to raise an error when not saved? As you may have guessed already,
```

> - So one the form objects need to raise an error when not saved? As you may have guessed already,

Reviewer 3: I can't quite grasp the meaning of the question. What do you want to convey to the reader here?

Author: It's not entirely a question, it's an informal way to say:
"Given one of the form objects need to raise an error when not saved, use the dangerous version of the method `save`"

Disclaimer: I missed an "of" at the beginning

Reviewer 3: OOOOHHH I see. Thanks

use the _dangerous_ `save!` method.
yaaf already defines it so that it raises an error just as any Active Record model does.
Of course, it will also trigger a database rollback when the models couldn't be saved.
Both `save` and `save!` methods are available.

> - Not sure where to put business logic related to the creation of an object? well, the form object is the place.

Reviewer 6: ```suggestion
- Not sure where to put business logic related to the creation of an object? Well, the form object is the place.
```

> yaaf will allow you to use callbacks the same way as your model callbacks work. For instance, if you want to send

Reviewer 2: ```suggestion
yaaf will allow you to use callbacks in the same way your model callbacks work. For instance, if you want to send
```

> an email after the form has been submited and persisted, the `after_commit` callback is the one you're looking for.

Reviewer 8: ```suggestion
an email after the form has been submitted and persisted, the `after_commit` callback is the one you're looking for.
```

More info can be found in the [readme](https://github.com/rootstrap/yaaf#callbacks).

> ### The Bob Ross of form objects

Reviewer 3: _Let's paint some happy little form objects over here_

Author: https://gfycat.com/excellentshorthammerheadshark

Author: I might be able to put that into this paragraph, what do you think?

Reviewer 3: 😆 that one is great!

> ### The Bob Ross of form objects

Reviewer: Benefit of Using Yet Another Active Form

> ### The Bob Ross of form objects

Reviewer 2: ```suggestion
### Benefit of Using Yet Another Active Form
```

Using yaaf will allow you to standardize your form object definitions across your project's production code.

> ### Why not use this?

Reviewer: Drawbacks of Yet Another Active Form

> ### Why not use this?

Reviewer 2: ```suggestion
### Drawbacks of Yet Another Active Form
```

If you would rather have controllers deal with all these responsibilities, then it's fine to keep doing it,
you may not need yaaf if that's the case.

Are you a service-objects-for-all person?
Then you might feel better writing service objects rather than form objects.

> If you work with Java, well you don't want to use this.

Reviewer 4: Maybe add `obviously don't` in this line 😆

Author: I'd rather leave it with a bit of sarcasm here

> If you work with Java, well you don't want to use this.

Reviewer 2: ```suggestion
If you work with Java, well you won't want to use this.
```

> ### A bit of history

Reviewer: The Background to Yet Another Active Form

> ### A bit of history

Reviewer 2: ```suggestion
### The Background to Yet Another Active Form
```

We have been using this very same approach on a production app for almost a year now,
more than 10 form objects written this way,
and it has served us well.
That's why we decided to extract it to a gem.
(yaaf is no more than [64](https://github.com/rootstrap/yaaf/blob/master/lib/yaaf/form.rb#L64) lines currently,
and we need no more)

Big thanks to [@santib](http://github.com/santib) for shaping up this gem from its very beginning.

> https://github.com/rootstrap/yaaf

Reviewer: For the original published post, visit [DEV Community](https://dev.to/juanmanuelramallo/yet-another-active-form-4874)

> https://github.com/rootstrap/yaaf

Reviewer 2: ```suggestion
https://github.com/rootstrap/yaaf 

For the original published post, visit [DEV Community](https://dev.to/juanmanuelramallo/yet-another-active-form-4874)
```

> https://github.com/rootstrap/yaaf

Reviewer 2: ```suggestion
For the original published post, visit [DEV Community](https://dev.to/juanmanuelramallo/yet-another-active-form-4874)
```
